globus_action_provider_tools/data_types.py

```diff
@@ -102,35 +107,111 @@ class ActionRequest(BaseModel):
+            "Action Providers may limit the maximum value. It is recommended "
+            "that Providers provide a default release_after value of approximately "
+            "30 days."
+        ),
     )
     monitor_by: Set[str] = Field(
         default_factory=set,
-        description="A list of principal URNs containing identities which are allowed to monitor the progress of the action using the status and log operations. When not provided, defaults to the user that initiated the action.",
-        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+        description=(
+            "A list of principal URNs containing identities which are allowed "
+            "to monitor the progress of the action using the status and log "
+            "operations. When not provided, defaults to the user that initiated "
+            "the action."
+        ),
+        regex=_principal_urn_regex,
     )
     manage_by: Set[str] = Field(
         default_factory=set,
-        description="A list of principal URNs containing identities which are allowed to manage the progress of the action using the cancel and release operations. When not provided, defaults to the user that initiated the action.",
-        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+        description=(
+            "A list of principal URNs containing identities which are allowed to "
+            "manage the progress of the action using the cancel and release "
+            "operations. When not provided, defaults to the user that initiated "
+            "the action."
+        ),
+        regex=_principal_urn_regex,
     )
     allowed_clients: List[str] = Field(
         default_factory=list, regex="^(public|globus|creator|.$)$"
     )
 
 
+# We provide some helper types for open dict fields like details. But, even when the
+# helper type is in place, it can be helpful to have the dict-like behavior to fall back
+# on. This MixIn provides support for some dict-like operations, but not all. Beware.
+class SubscriptableObject:
+    def __getitem__(self, key):
+        return vars(self).get(key)
+
+    def __len__(self):
+        return len(vars(self))
+
+    def __iter__(self):
+        return vars(self).keys()
+
+    def pop(self, key, *args):
+        if len(args) > 0:
+            return vars(self).pop(key, args[0])
+        else:
+            return vars(self).pop(key)
+
+
+class ExtensibleCodeDescription(BaseModel, SubscriptableObject):
+    class Config:
+        extra = "allow"
+
+    code: str = Field(..., description=(""))
+    description: str = Field(..., description=(""))
+
+
+class ActionFailedDetails(ExtensibleCodeDescription):
+    pass
+
+
+class PaginationWrapper(BaseModel):
+    limit: int
+    has_next_page: bool
+    marker: Optional[str]
+
+
+class ActionLogEntry(ExtensibleCodeDescription):
```

Comment:
I guess we could default to the `GenericLogEntry` code here. That might make it a little easier to construct these, but it could be that you were thinking there'd be different types of log entries, in which case I think this is good.

Reply:
Yeah, I don't know about defaulting to that. On the one hand, we don't really specify what the codes should/could be. On the other hand, I kinda like people who are actually gonna do logs (probably a small subset) to think about what they want there.

Reply:
> I kinda like people who are actually gonna do logs

FTFY :)